superset/utils/webdriver.py

```diff
@@ -56,67 +67,184 @@ class ChartStandaloneMode(Enum):
+            logger.debug(
+                "%i alert elements have been found in the screenshot", len(alert_divs)
+            )
 
-            # wait until the modal becomes invisible
-            WebDriverWait(
-                driver, current_app.config["SCREENSHOT_WAIT_FOR_ERROR_MODAL_INVISIBLE"]
-            ).until(EC.invisibility_of_element(modal))
+            for alert_div in alert_divs:
+                # See More button
+                alert_div.get_by_role("button").click()
+
+                # wait for modal to show up
+                page.wait_for_selector(
+                    ".ant-modal-content",
+                    timeout=current_app.config[
+                        "SCREENSHOT_WAIT_FOR_ERROR_MODAL_VISIBLE"
+                    ]
+                    * 1000,
+                    state="visible",
+                )
+                err_msg_div = page.locator(".ant-modal-content .ant-modal-body")
+                #
+                # # collect error message
+                error_messages.append(err_msg_div.text_content())
+                #
+                # # Use HTML so that error messages are shown in the same style (color)
+                error_as_html = err_msg_div.inner_html().replace("'", "\\'")
+                #
+                # # close modal after collecting error messages
+                page.locator(".ant-modal-content .ant-modal-close").click()
+                #
+                # # wait until the modal becomes invisible
+                page.wait_for_selector(
+                    ".ant-modal-content",
+                    timeout=current_app.config[
+                        "SCREENSHOT_WAIT_FOR_ERROR_MODAL_INVISIBLE"
+                    ]
+                    * 1000,
+                    state="detached",
+                )
+                try:
+                    # Even if some errors can't be updated in the screenshot,
+                    # keep all the errors in the server log and do not fail the loop
+                    alert_div.evaluate(
+                        "(node, error_html) => node.innerHtml = error_html",
+                        [error_as_html],
+                    )
+                except Error:
+                    logger.exception("Failed to update error messages using alert_div")
+        except Error:
+            logger.exception("Failed to capture unexpected errors")
 
-            # Use HTML so that error messages are shown in the same style (color)
-            error_as_html = err_msg_div.get_attribute("innerHTML").replace("'", "\\'")
+        return error_messages
 
+    def get_screenshot(self, url: str, element_name: str, user: User) -> bytes | None:
+        with sync_playwright() as playwright:
+            browser = playwright.chromium.launch()
+            context = browser.new_context(
+                bypass_csp=True,
+                viewport={
+                    "height": self._window[1],
+                    "width": self._window[0],
```

Comment:
we have a new feature to allow users to select the viewport width, and also some configs for selenium to set the pixel ratio for the image (for higher resolution images). Do we have compatibility here in playwright with those two features?

Reply:
Custom viewport width is supported. I added the pixel ratio support, thanks for spotting that!